modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/config/DynamicConfigurationInstanceHolder.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.extension.internal.runtime.config;
+
+import org.mule.extension.runtime.Expirable;
+import org.mule.extension.runtime.ExpirationPolicy;
+
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * A holder class which decorates a {@link #configurationInstance} with extra
+ * management information such as registration name, usage tracking, etc.
+ * <p/>
+ * This class is thread safe.
+ *
+ * @since 4.0
+ */
+final class DynamicConfigurationInstanceHolder implements Expirable
+{
+
+    private final String registrationName;
+    private final Object configurationInstance;
+
+    private final AtomicInteger usageCount = new AtomicInteger(0);
+    private long lastUsedMillis = now();
+
+    public DynamicConfigurationInstanceHolder(String registrationName, Object configurationInstance)
+    {
+        this.registrationName = registrationName;
+        this.configurationInstance = configurationInstance;
+    }
+
+    /**
+     * Determines if the held {@link #configurationInstance} should be expired per
+     * the given {@code expirationPolicy}.
+     * <p/>
+     * The first thing this method checks is if there's any ongoing operation using the held
+     * configuration instance. If that's not the case, then it compares the timestamp of when
+     * the {@link #configurationInstance} was last used against the {@code expirationPolicy}
+     *
+     * @param expirationPolicy a {@link ExpirationPolicy}
+     * @return {@code true} if the {@link #configurationInstance} should be expired. {@code false} otherwise
+     */
+    @Override
+    public boolean isExpired(ExpirationPolicy expirationPolicy)
+    {
+        return isNotInUse() && expirationPolicy.isExpired(lastUsedMillis, TimeUnit.MILLISECONDS);
+    }
+
+    /**
+     * Returns the held configuration instance
+     */
+    public Object getConfigurationInstance()
+    {
+        return configurationInstance;
+    }
+
+    /**
+     * acknowledges and tracks the fact that there's an ongoing operation execution
+     * using the {@link #configurationInstance}. This method is thread safe.
+     *
+     * @return the updated usage count.
+     */
+    public int accountUsage()
```

Comment:
don't agree with this one. It doesn't start with the get prefix, so that should be hint enough that this isn't just a getter.
Also, the account verb and the javadoc are pretty explicit on this method actually modifying state in order to track the usage.

Finally, this method does return the current number of usages, so what the method suggests you is what it actually does, which is good